dokan/loaddll.go

```diff
@@ -0,0 +1,144 @@
+// Copyright 2018 Keybase Inc. All rights reserved.
+// Use of this source code is governed by a BSD
+// license that can be found in the LICENSE file.
+
+// +build windows
+
+package dokan
+
+/*
+#include "bridge.h"
+*/
+import "C"
+
+import (
+	"bytes"
+	"crypto/sha256"
+	"fmt"
+	"io"
+	"os"
+	"runtime"
+	"strings"
+	"unsafe"
+
+	"golang.org/x/sys/windows"
+)
+
+const shortPath = `DOKAN1.DLL`
+const syswow64 = `C:\WINDOWS\SYSWOW64\`
+const system32 = `C:\WINDOWS\SYSTEM32\`
+
+type errorPrinter struct {
+	buf bytes.Buffer
+}
+
+func (ep *errorPrinter) Printf(s string, os ...interface{}) {
+	fmt.Fprintf(&ep.buf, s, os...)
+}
+
+// loadLibrary calls win32 LoadLibrary and logs the result.
+func loadLibrary(epc *errorPrinter, path string) (windows.Handle, error) {
+	hdl, err := windows.LoadLibrary(path)
+	epc.Printf("LoadLibrary(%q) -> %v,%v", path, hdl, err)
+	return hdl, err
+}
+
+// doLoadDLL tries to load the dokan DLL from various locations.
+func doLoadDLL(epc *errorPrinter, path string) (windows.Handle, error) {
+	var guessPath bool
+	epc.Printf("loadDokanDLL %q", path)
+	if path == "" {
+		path = shortPath
+		guessPath = true
+	} else {
+		path = strings.Replace(path, `/`, `\`, -1)
+	}
+	const loadLibrarySearchSystem32 = 0x800
+	const flags = loadLibrarySearchSystem32
+	hdl, err := windows.LoadLibraryEx(path, 0, flags)
+	epc.Printf("loadDokanDLL LoadLibraryEx(%q,0,%x) -> %v,%v", path, flags, hdl, err)
+	if err == nil || !guessPath {
+		return hdl, err
+	}
+	// User probably has not installed KB2533623 which is a security update
+	// from 2011. Without this Windows security update loading libraries
+	// is unsafe on Windows.
+
+	if runtime.GOARCH == `386` {
```

Comment:
runtime.GOARCH is set at compile time and is unrelated to the architecture the program is running on. I assume you wanted that? The updater gets it this way: https://github.com/keybase/go-updater/blob/master/keybase/platform_windows.go#L127

Reply:
Actually doing it by the compile time version is fine in this case. If it was compiled as 32-bit then we want:
+ 32-bit windows: syswow64 does not exist, system32 exists and gets used.
+ 64-bit windows: syswow64 exists and gets used instead of system32.
The old C version also worked like this (but it just checked sizeof(void*)),

Reply:
ah ok then, NVM!